src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,196 @@
+package duke;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Represent the file to store the task list.
+ */
+public class Storage {
+    private final String filePath;
+    private final Path path;
+
+    /**
+     * Constructor to create a storage.
+     *
+     * @param filepath string representation of the path.
+     * @throws DukeFileNotFoundException indicate attempt to open the file denoted by a specified pathname has failed.
+     */
+    public Storage(String filepath) throws DukeFileNotFoundException {
+        try {
+            this.filePath = filepath;
+            this.path = Path.of(filepath);
+        } catch (InvalidPathException e) {
+            throw new DukeFileNotFoundException("File is not found!");
+        }
+    }
+
+    /**
+     * Set a default storage if the file or folder is not found.
+     */
+    protected static void setDefaultStorage() {
+        File folder = new File("data/");
+        if (!folder.exists()) {
+            folder.mkdir();
+        }
+
+        File file = new File("data/duke.txt");
+        if (!file.exists()) {
+            try {
+                file.createNewFile();
+            } catch (IOException e) {
+                System.out.println("IO Exception occur");
+            }
+        }
+    }
+
+    /**
+     * Insert new task into the file.
+     *
+     * @param t task to be inserted.
+     * @throws DukeIoException indicate failed or interrupted I/O operations occurred.
+     */
+    protected void updateData(Task t) throws DukeIoException {
+        try {
+            List<String> allLine = Files.readAllLines(path);
+            String s = t.toString().charAt(1) + " | " + t.getStatusIcon() + " | " + t.description;
+            if (t instanceof Deadlines) {
+                Deadlines d = (Deadlines) t;
+                s += " | " + d.getBy();
+            } else if (t instanceof Events) {
+                Events e = (Events) t;
+                s += " | " + e.getStart() + " | " + e.getEnd();
+            }
+            allLine.add(s);
+            Files.write(path, allLine);
+        } catch (IOException e) {
+            throw new DukeIoException("Cannot read from " + filePath + " data file");
+        }
+    }
+
+    /**
+     * Update the task in the file with its new state.
+     *
+     * @param lineNumber line number in the file to be toggled.
+     * @param status status of file that indicate whether the task is done or not.
+     * @throws DukeIoException indicate failed or interrupted I/O operations occurred.
+     */
+    protected void updateData(int lineNumber, int status) throws DukeIoException {
+        try {
+            List<String> allLine = Files.readAllLines(path);
```

Comment:
Plural form of variable "allLine" can be considered

Reply:
Nice catch! thanks